lib/ansible/template/__init__.py

```diff
@@ -252,7 +252,19 @@ def _unroll_iterator(func):
     explicitly use ``|list`` to unroll.
     """
     def wrapper(*args, **kwargs):
-        ret = func(*args, **kwargs)
+        try:
+            ret = func(*args, **kwargs)
+        except (UndefinedError, AnsibleUndefinedVariable):
+            raise
+        except Exception as e:
+            if 'undefined' in to_native(e):
```

Comment:
I'm not sure if this is a good idea, but I noticed that lookups don't have the same problem since we check that the terms are defined first and thought maybe we could retroactively check for filters. @nitzmahone What do you think?

Reply:
We have another wrapper that intercepts the result of a filter: `_wrap_native_text` so we would need to handle that there too, or create an additional wrapper just for this purpose. Also should we do the same for jinja tests?

Reply:
That was the thing I was saying earlier- checking for/failing on undefined inputs upfront would probably be what most people would expect, but that'd be a breaking change. Even if we chose that breaking change, we'd probably need a way for filters that *want* to allow undefined inputs to do so (transparently to the caller of the filter).

And yeah, we probably need an analogue for tests, since the same problems exist.

Reply:
I'm also unclear on why this needs to be here- the string check for `undefined` in the exception message feels dirty and a potential source of false-trip. What are we getting that's not in the `Undefined` class hierarchy- is it something we can fix elsewhere to get rid of this?

Reply:
@mkrizek Thanks for your insight!

@nitzmahone Yeah, I see what you were saying and that seems very featureish.

> I'm also unclear on why this needs to be here- the string check for undefined in the exception message feels dirty and a potential source of false-trip. What are we getting that's not in the Undefined class hierarchy- is it something we can fix elsewhere to get rid of this?

This (36db9950af4447b0d0c33ffc683c2d42c698d752) was the continuation of my thoughts on trying to fix 3rd party plugins. I kept the commits separate so it would be easy to drop if we don't want it - I just wanted to put it out there since `except Exception:` is common and was a suggested pattern in our plugin dev documentation.

Reply:
Another option might be to build it the way we wish it were, then start a deprecation warning cycle on this bridging behavior (eg, do something like this and/or input inspection with an explicit optin/optout in plugin metadata, decorator, or whatever, then issue a dep warning on the exception catch or undefined inputs for plugins that haven't explicitly opted into receiving them).

:shrug:

Reply:
I had another idea for the 'undefined' hack, and I think it's in better shape now :crossed_fingers: